### ISSUE TITLE
Stuck on "Lutris is starting" on Ubuntu 20.10

### ISSUE THREAD

**Issue by atrauzzi** [lutris/lutris#3496]
Fresh install of Ubuntu 20.10.

```
2021-04-06 07:36:41,432: Download failed: ffi_prep_closure(): bad user_data (it seems that the version of the libffi library seen at runtime is different from the 'ffi.h' file seen at compile-time)
```

I've done nothing special to my system with regards to python or python libraries.  This seems like a recurring issue, is there any chance that Lutris needs to bundle this dependency somehow, or update to support the newer version(s)?

**Comment by vanhoxx** [comment]
Same problem on me. First I saw this on Pop!_OS 20.10 then I fresh installed Linux Mint 20.1 and problem still the same. I guess it's about Ubuntu, maybe there is a missing dependency with repos or not supporting the newest version.

**Comment by Cattodeveloper909** [comment]
You can open lutris but it stays in the "Lutris is starting" box right? I had that problem on Pop!_OS 20.10 as @vanhoxx said, what I did is just to right click on the app and close it, then it just started, I know that this isn't a fix but it can help, and I imagine that you are installing for the PPA right? Hope I could help you in some way

**Comment by atrauzzi** [comment]
That doesn't actually result in the application working.  It's still throwing errors in the background trying to use mismatched python libs.

**Comment by Cattodeveloper909** [comment]
Ohhhh okay I get the error know, is python, well that explains it, python had a recent big update to 3.9, and the majority of apps that uses python are making the switch, as lutris, I would say to try to update it, uninstall lutris and all of their dependencies, and install it again and see how it goes! you can see other similar problems like this #3504

**Comment by atrauzzi** [comment]
That's not the issue.  The issue is that Lutris is depending on a library of a specific version that has since been updated.

Uninstalling and reinstalling Python doesn't fix the issue.

**Comment by Morodar** [comment]
Pop!_OS and Mint both are based on Ubuntu.

The exception occurs in `OpenSSL/SSL.py`
```
 File "/usr/lib/python3/dist-packages/OpenSSL/SSL.py", line 337, in __init__
    self.callback = _ffi.callback(

 Download failed: ffi_prep_closure(): bad user_data (it seems that the version of the libffi library seen at runtime is different from the 'ffi.h' file seen at compile-time)

```

This bug has also been reported in the pyopenssl project:  
https://bugs.launchpad.net/ubuntu/+source/pyopenssl/+bug/1903890

**Comment by Morodar** [comment]
After upgrading to Ubuntu 21.04, `libffi7` has been installed during the update and Lutris is working again for me!

**Comment by vanhoxx** [comment]
https://packages.ubuntu.com/focal/libffi7 
In Ubuntu 20.04 LTS version, libffi7 is available in repository and I already have this package. Problem still the same. Maybe there is a incompatibility in 20.10.

**Comment by kajlando** [comment]
python3.9 is attempting to resolve lutris.net (internet access)
When I block it the message "Lutris is starting" appears, I have to close it via panel (Plank) and even if the main window has already opened it still continues for some seconds.

**Comment by kajlando** [comment]
python3.9 is attempting to resolve lutris.net (internet access)
When I block it the message "Lutris is starting" appears, I have to close it via panel (Plank) and even if the main window has already opened it still continues for some seconds.
So try Plank->right click>close->main windows should appear

**Comment by kajlando** [comment]
python3.9 is attempting to resolve lutris.net (internet access)
When I block it the message "Lutris is starting" appears, I have to close it via panel (Plank) and even if the main window has already opened it still continues for some seconds.
So try Plank->right click->close->main windows should appear.

**Comment by kajlando** [comment]
python3.9 is attempting to resolve lutris.net (internet access).
When I block it the message "Lutris is starting" appears, I have to close it via panel (Plank) and even if the main window has already opened it still continues for some seconds.
So try Plank->right click->close->main windows should appear.

**Comment by kajlando** [comment]
python3.9 is attempting to resolve lutris.net (internet access).
When I block it the message "Lutris is starting" appears, I have to close it via panel (Plank) and even if the main window has already opened it still continues for some seconds.
So try Plank->right click->close->main window should appear.

**Comment by kajlando** [comment]
python3.9 is attempting to resolve lutris.net (internet access).
When I block it the message "Lutris is starting" appears, I have to close it via panel (Plank) and even if the main window has already opened it still continues for some seconds.

**Comment by Jozro** [comment]
The problem is caused by a new actualization, I hve that problen in manjaro, arch and ubuntu, but I installed ubuntu gamepack, lutris is included there, lutris run perfect, and I could play, but when I installed updates for ubuntu gamepack, the bug regressed.
the problem is in the updates.

**Comment by promil23** [comment]
I found the issue
```
Python 3.9.5 (default, May 11 2021, 08:20:37) 
[GCC 10.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import urllib.request
>>> r = urllib.request.urlopen('https://lutris.net')
('2606:4700:3031::ac43:bd13', 443, 0, 0)
^CTraceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.9/urllib/request.py", line 214, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.9/urllib/request.py", line 517, in open
    response = self._open(req, data)
  File "/usr/lib/python3.9/urllib/request.py", line 534, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
  File "/usr/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.9/urllib/request.py", line 1389, in https_open
    return self.do_open(http.client.HTTPSConnection, req,
  File "/usr/lib/python3.9/urllib/request.py", line 1346, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/usr/lib/python3.9/http/client.py", line 1253, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.9/http/client.py", line 1299, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.9/http/client.py", line 1248, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.9/http/client.py", line 1008, in _send_output
    self.send(msg)
  File "/usr/lib/python3.9/http/client.py", line 948, in send
    self.connect()
  File "/usr/lib/python3.9/http/client.py", line 1415, in connect
    super().connect()
  File "/usr/lib/python3.9/http/client.py", line 919, in connect
    self.sock = self._create_connection(
  File "/usr/lib/python3.9/socket.py", line 832, in create_connection
    sock.connect(sa)
KeyboardInterrupt
```

To get **lutris.net** website Lutris uses urllib.urlopen which at the lowest level uses python's http.client library. 
I put print statement before ```sock.connect(sa)``` to see what it's trying to connect to and it turned out it tries to connect to **IPv6 address !** ```('2606:4700:3031::ac43:bd13', 443, 0, 0)``` which stucks indefinitely. 

So **turning off IPv6** fixed the issue and Lutris starts and works fine again !
```
sudo sysctl -w net.ipv6.conf.all.disable_ipv6=1
sudo sysctl -w net.ipv6.conf.default.disable_ipv6=1
sudo sysctl -w net.ipv6.conf.lo.disable_ipv6=1
```

```
>>> r = urllib.request.urlopen('https://lutris.net')
('172.67.189.19', 443)
```
PS. I've tested it on Ubuntu 21.04, kernel 5.11.0-34

**Comment by fzlrog** [comment]
> So **turning off IPv6** fixed the issue and Lutris starts and works fine again !
> 
> ```
> sudo sysctl -w net.ipv6.conf.all.disable_ipv6=1
> sudo sysctl -w net.ipv6.conf.default.disable_ipv6=1
> sudo sysctl -w net.ipv6.conf.lo.disable_ipv6=1
> ```
> 
> ```
> >>> r = urllib.request.urlopen('https://lutris.net')
> ('172.67.189.19', 443)
> ```
> 
> PS. I've tested it on Ubuntu 21.04, kernel 5.11.0-34

THANK YOU! I've been wracking my brain trying to get Lutris to work and this fixed it for me.

**Comment by RandamuMaki** [comment]
I also had this issue today on Debian Bullseye with Lutris 0.5.8.3 as available in the Debian repositories.
Can someone please add a IPv4 backup in case of IPv6 timeout? Or just rely on DNS for the IPv4/IPv6 reroute? I'd rather not turn IPv6 off system-wide just for Lutris.

**Comment by Leopard1907** [comment]
Same problem here as well. Solved by disabling ipv6 as mentioned above for multiple times.

**Comment by ryannathans** [comment]
Taking forever when checking for runtime updates when IPv6 was blocked by VPN software on my PC - turned that software off (fixing IPv6), restarted lutris, and it started like normal.

**Comment by opidopiopi** [comment]
If you don't want to disable ipv6 on your system you can also uncomment the line `precedence ::ffff:0:0/96 100` in /etc/gai.conf as mentioned in [this](https://github.com/googleapis/google-api-python-client/issues/709#issuecomment-690942635) thread. That works for me.

**Comment by strycore** [comment]
fixed

**Comment by runboy93** [comment]
> Taking forever when checking for runtime updates when IPv6 was blocked by VPN software on my PC - turned that software off (fixing IPv6), restarted lutris, and it started like normal.

I was other way, when used my connection without VPN with latest 0.5.10 beta1, download took 10 minutes (until crashed due error with coverart, reported on bugs already), and when tried whole thing again from start with VPN it only took couple of minutes.

**Comment by DocBretti** [comment]
> I found the issue
> 
> ```
> Python 3.9.5 (default, May 11 2021, 08:20:37) 
> [GCC 10.3.0] on linux
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import urllib.request
> >>> r = urllib.request.urlopen('https://lutris.net')
> ('2606:4700:3031::ac43:bd13', 443, 0, 0)
> ^CTraceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/usr/lib/python3.9/urllib/request.py", line 214, in urlopen
>     return opener.open(url, data, timeout)
>   File "/usr/lib/python3.9/urllib/request.py", line 517, in open
>     response = self._open(req, data)
>   File "/usr/lib/python3.9/urllib/request.py", line 534, in _open
>     result = self._call_chain(self.handle_open, protocol, protocol +
>   File "/usr/lib/python3.9/urllib/request.py", line 494, in _call_chain
>     result = func(*args)
>   File "/usr/lib/python3.9/urllib/request.py", line 1389, in https_open
>     return self.do_open(http.client.HTTPSConnection, req,
>   File "/usr/lib/python3.9/urllib/request.py", line 1346, in do_open
>     h.request(req.get_method(), req.selector, req.data, headers,
>   File "/usr/lib/python3.9/http/client.py", line 1253, in request
>     self._send_request(method, url, body, headers, encode_chunked)
>   File "/usr/lib/python3.9/http/client.py", line 1299, in _send_request
>     self.endheaders(body, encode_chunked=encode_chunked)
>   File "/usr/lib/python3.9/http/client.py", line 1248, in endheaders
>     self._send_output(message_body, encode_chunked=encode_chunked)
>   File "/usr/lib/python3.9/http/client.py", line 1008, in _send_output
>     self.send(msg)
>   File "/usr/lib/python3.9/http/client.py", line 948, in send
>     self.connect()
>   File "/usr/lib/python3.9/http/client.py", line 1415, in connect
>     super().connect()
>   File "/usr/lib/python3.9/http/client.py", line 919, in connect
>     self.sock = self._create_connection(
>   File "/usr/lib/python3.9/socket.py", line 832, in create_connection
>     sock.connect(sa)
> KeyboardInterrupt
> ```
> 
> To get **lutris.net** website Lutris uses urllib.urlopen which at the lowest level uses python's http.client library. I put print statement before `sock.connect(sa)` to see what it's trying to connect to and it turned out it tries to connect to **IPv6 address !** `('2606:4700:3031::ac43:bd13', 443, 0, 0)` which makes it stuck.
> 
> So **turning off IPv6** fixed the issue and Lutris starts and works fine again !
> 
> ```
> sudo sysctl -w net.ipv6.conf.all.disable_ipv6=1
> sudo sysctl -w net.ipv6.conf.default.disable_ipv6=1
> sudo sysctl -w net.ipv6.conf.lo.disable_ipv6=1
> ```
> 
> ```
> >>> r = urllib.request.urlopen('https://lutris.net')
> ('172.67.189.19', 443)
> ```
> 
> PS. I've tested it on Ubuntu 21.04, kernel 5.11.0-34

Thnx mate, this worked on garuda, too. It may be a dirty solution but for now it works. :-)

**Comment by cheater** [comment]
It's not actually _stuck_, it's just taking its sweet-ass time to time out:

"The optional `timeout` parameter specifies a timeout in seconds for blocking operations like the connection attempt (if not specified, the **global default timeout setting** will be used)."

Please set it to 10 seconds, anything higher than that is too long for a splash screen anyways. Eventually, this should be done in the background _after_ start-up like people have complained [elsewhere](https://www.reddit.com/r/linux_gaming/comments/r122ue/comment/hlw9kqk/?utm_source=reddit&utm_medium=web2x&context=3). There should be no splash screens, auto-updates, or any other wait dialogs when launching a launcher - from a usability perspective it's "bad enough" that you have to click through to whatever you actually want.

**Comment by cheater** [comment]
In my case, this came about because for some reason on my network lutris.net resolves to 188.114.96.12, which is unreachable, whereas it seems like for most people it should resolve to 104.21.73.76, which is reachable.

**Comment by danieljohnson2** [comment]
For me it resolves to 172.67.189.19. Content network shenanigans I guess.

I find I can ping 188.114.96.12 though. Not sure why you cannot.

**Comment by cheater** [comment]
@danieljohnson2 188.114.96.12 is probably limited to responding to certain geographical regions; discussion continues in #4450
